File: DA/python/load_revenue_to_warehouse.py
"""
Load revenue raw data (JSON) into the warehouse.
Events under "events"; amount sometimes string; dates mixed; content_id/campaign_id may be null, missing, or "".

Run from project root: py python/load_revenue_to_warehouse.py
"""

import json
import sqlite3
import re
from pathlib import Path

PROJECT_ROOT = Path(__file__).resolve().parent.parent
RAW_REVENUE = PROJECT_ROOT / "raw" / "api" / "revenue_2024q1.json"
WAREHOUSE_DB = PROJECT_ROOT / "warehouse" / "brightwave.sqlite"
SCHEMA_SQL = PROJECT_ROOT / "warehouse" / "schema.sql"
SEED_SQL = PROJECT_ROOT / "warehouse" / "seed_dimensions.sql"
SEED_CAMPAIGNS_SQL = PROJECT_ROOT / "warehouse" / "seed_campaigns.sql"


def normalize_date(value):
    if value is None or (isinstance(value, str) and value.strip() == ""):
        return None
    s = str(value).strip()
    m = re.match(r"^(\d{4})-(\d{2})-(\d{2})", s)
    if m:
        return f"{m.group(1)}-{m.group(2)}-{m.group(3)}"
    m = re.match(r"^(\d{1,2})/(\d{1,2})/(\d{4})$", s)
    if m:
        mo, day, yr = int(m.group(1)), int(m.group(2)), int(m.group(3))
        return f"{yr}-{mo:02d}-{day:02d}"
    return None


def safe_float(value):
    if value is None:
        return None
    if isinstance(value, (int, float)):
        return float(value)
    s = str(value).strip()
    if s == "" or s.lower() == "null":
        return None
    try:
        return float(s)
    except (ValueError, TypeError):
        return None


def optional_id(value):
    """Return stripped string or None for FK columns (content_id, campaign_id)."""
    if value is None:
        return None
    s = str(value).strip()
    return s if s else None


def run_sql_file(conn, path):
    with open(path, "r", encoding="utf-8") as f:
        conn.executescript(f.read())


def main():
    WAREHOUSE_DB.parent.mkdir(parents=True, exist_ok=True)
    conn = sqlite3.connect(WAREHOUSE_DB)

    run_sql_file(conn, SCHEMA_SQL)
    run_sql_file(conn, SEED_SQL)
    run_sql_file(conn, SEED_CAMPAIGNS_SQL)
    conn.commit()

    with open(RAW_REVENUE, "r", encoding="utf-8") as f:
        raw = json.load(f)
    events = raw.get("events", [])

    cleaned = []
    for e in events:
        event_id = (e.get("id") or "").strip()
        if not event_id:
            continue
        date_iso = normalize_date(e.get("date"))
        if not date_iso:
            continue
        amount = safe_float(e.get("amount"))
        if amount is None:
            continue
        etype = (e.get("type") or "").strip()
        if not etype:
            continue
        content_id = optional_id(e.get("content_id"))
        campaign_id = optional_id(e.get("campaign_id"))
        cleaned.append({
            "event_id": event_id,
            "content_id": content_id,
            "campaign_id": campaign_id,
            "date": date_iso,
            "amount": amount,
            "type": etype,
        })

    cur = conn.cursor()
    for row in cleaned:
        cur.execute(
            """
            INSERT OR REPLACE INTO fact_revenue_events
            (event_id, content_id, campaign_id, date, amount, type)
            VALUES (?, ?, ?, ?, ?, ?)
            """,
            (
                row["event_id"],
                row["content_id"],
                row["campaign_id"],
                row["date"],
                row["amount"],
                row["type"],
            ),
        )
    conn.commit()
    conn.close()
    print(f"Loaded {len(cleaned)} revenue events into fact_revenue_events.")


if __name__ == "__main__":
    main()

File: DA/r/load_revenue_to_warehouse.R
# Load revenue raw data (JSON) into the warehouse.
# Events under "events"; amount sometimes string; content_id/campaign_id may be missing or "".
# Required: DBI, RSQLite, jsonlite
# Run from project root: Rscript r/load_revenue_to_warehouse.R

library(DBI)
library(RSQLite)
library(jsonlite)

`%||%` <- function(a, b) if (!is.null(a)) a else b

project_root <- getwd()
if (basename(project_root) == "r") project_root <- dirname(project_root)
raw_revenue <- file.path(project_root, "raw", "api", "revenue_2024q1.json")
warehouse_db <- file.path(project_root, "warehouse", "brightwave.sqlite")
schema_sql <- file.path(project_root, "warehouse", "schema.sql")
seed_sql <- file.path(project_root, "warehouse", "seed_dimensions.sql")
seed_campaigns_sql <- file.path(project_root, "warehouse", "seed_campaigns.sql")

run_sql_file <- function(conn, path) {
  sql <- paste(readLines(path, encoding = "UTF-8"), collapse = "\n")
  statements <- strsplit(sql, ";\\s*\n")[[1]]
  for (s in statements) {
    s <- trimws(s)
    if (nchar(s) > 0) dbExecute(conn, s)
  }
}

normalize_date <- function(value) {
  if (is.null(value) || (is.character(value) && trimws(value) == ""))
    return(NA_character_)
  s <- trimws(as.character(value))
  if (grepl("^[0-9]{4}-[0-9]{2}-[0-9]{2}", s)) return(substr(s, 1, 10))
  d <- as.Date(s, format = "%m/%d/%Y")
  if (!is.na(d)) return(format(d, "%Y-%m-%d"))
  NA_character_
}

safe_float <- function(x) {
  if (is.null(x)) return(NA_real_)
  if (is.character(x) && trimws(x) %in% c("", "null", "NULL")) return(NA_real_)
  as.numeric(x)
}

optional_id <- function(x) {
  if (is.null(x)) return(NA_character_)
  s <- trimws(as.character(x))
  if (nchar(s) == 0) return(NA_character_)
  s
}

# ---- Main ----

dir.create(dirname(warehouse_db), recursive = TRUE, showWarnings = FALSE)
conn <- dbConnect(RSQLite::SQLite(), warehouse_db)

run_sql_file(conn, schema_sql)
run_sql_file(conn, seed_sql)
run_sql_file(conn, seed_campaigns_sql)

raw <- read_json(raw_revenue)
events <- raw$events

event_id   <- vapply(events, function(e) trimws(e$id %||% ""), character(1))
date_iso   <- vapply(events, function(e) normalize_date(e$date), character(1))
amount     <- vapply(events, function(e) safe_float(e$amount), numeric(1))
etype      <- vapply(events, function(e) trimws(e$type %||% ""), character(1))
content_id <- vapply(events, function(e) optional_id(e$content_id), character(1))
campaign_id <- vapply(events, function(e) optional_id(e$campaign_id), character(1))

# REV-011 has no campaign_id key: jsonlite gives NULL, optional_id -> NA_character_
# Replace "" with NA for DB NULL
content_id[content_id == ""] <- NA_character_
campaign_id[campaign_id == ""] <- NA_character_

keep <- nchar(event_id) > 0 & !is.na(date_iso) & !is.na(amount) & nchar(etype) > 0
event_id   <- event_id[keep]
content_id <- content_id[keep]
campaign_id <- campaign_id[keep]
date_iso   <- date_iso[keep]
amount     <- amount[keep]
etype      <- etype[keep]

for (i in seq_along(event_id)) {
  dbExecute(conn,
    "INSERT OR REPLACE INTO fact_revenue_events (event_id, content_id, campaign_id, date, amount, type) VALUES (?, ?, ?, ?, ?, ?)",
    params = list(event_id[i], content_id[i], campaign_id[i], date_iso[i], amount[i], etype[i]))
}

dbDisconnect(conn)
message("Loaded ", sum(keep), " revenue events into fact_revenue_events.")
